### ISSUE TITLE
Refund API V2 Error Details

### ISSUE THREAD

**Issue by marcmoy** [recurly/recurly-client-ruby#652]
Hi, is there a way to get additional details about a `#<Recurly::API::UnprocessableEntity: 422 Unprocessable Entity (POST https://api.recurly.com/v2/invoices/..../refund)` error? The `Recurly::API::UnprocessableEntity` class does not provide a method to inspect the exact reason behind the error. Based on what I'm seeing in the Recurly dashboard, I'm noticing the refund could be rejected due to an existing chargeback, bank decline, or card stolen error. We would like to perform different actions in each case.

(Using release version `2.17.11`)

**Comment by douglasmiller** [comment]
Hey @marcmoy, 

The `UnprocessableEntity` error extends `ClientError` which extends `ResponseError`. You should be able to obtain adequate information about the cause of the error from [the methods on `ResponseError`](https://github.com/recurly/recurly-client-ruby/blob/v2/lib/recurly/api/errors.rb#L6). Have you checked `UnprocessableEntity#symbol`, `UnprocessableEntity#description`, and/or `UnprocessableEntity#details`?

**Comment by marcmoy** [comment]
@douglasmiller Yes I tried inspecting the error but not seeing the results I'm expecting:
```
irb(main):032:0> e
=> #<Recurly::API::UnprocessableEntity: 422 Unprocessable Entity (POST https://api.recurly.com/v2/invoices/114677553/refund)>
irb(main):033:0> e.symbol
=> nil
irb(main):034:0> e.description
=> nil
irb(main):035:0> e.details
=> nil
irb(main):036:0> e.message
=> "422 Unprocessable Entity (POST https://api.recurly.com/v2/invoices/114677553/refund)"
irb(main):037:0> e.methods
=> [:to_s, :details, :description, :code, :symbol, :request, :response, :set_message, :==, :respond_to?, :inspect, :exception, :message, :backtrace, :backtrace_locations, :set_backtrace, :cause, :to_yaml, :taguri, :taguri=, :blame_file!, :blamed_files, :describe_blame, :copy_blame!, :`, :to_yaml_style, :to_yaml_properties, :blank?, :present?, :presence, :deep_dup, :acts_like?, :duplicable?, :try, :try!, :in?, :presence_in, :to_param, :to_query, :as_json, :instance_values, :instance_variable_names, :to_json_with_active_support_encoder, :to_json_without_active_support_encoder, :to_json, :with_options, :html_safe?, :psych_to_yaml, :dclone, :syck_to_yaml, :pretty_print, :pretty_print_cycle, :pretty_print_instance_variables, :pretty_print_inspect, :require_dependency, :unloadable, :require_or_load, :load_dependency, :tap, :public_send, :instance_variables, :instance_variable_set, :instance_variable_defined?, :remove_instance_variable, :private_methods, :kind_of?, :is_a?, :instance_variable_get, :method, :public_method, :singleton_method, :instance_of?, :class_eval, :silence_stderr, :silence_stream, :extend, :with_warnings, :quietly, :define_singleton_method, :capture, :enable_warnings, :suppress, :to_enum, :enum_for, :suppress_warnings, :silence, :concern, :<=>, :===, :=~, :!~, :eql?, :freeze, :display, :object_id, :send, :pretty_inspect, :gem, :silence_warnings, :nil?, :hash, :class, :singleton_class, :clone, :dup, :itself, :taint, :tainted?, :untaint, :untrust, :trust, :untrusted?, :methods, :protected_methods, :frozen?, :public_methods, :singleton_methods, :!, :!=, :__send__, :equal?, :instance_eval, :instance_exec, :__id__]
```

**Comment by marcmoy** [comment]
Following up... I just discovered the raw xml response is extractable via `e.response.body`.
```
irb(main):057:0> e.response.body
<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n
<errors>\n
<transaction_error>\n
<error_code>deposit_referenced_chargeback</error_code>....
```
However, I'm curious if there is a method available to return the `transaction_error` instead of having to parse the XML response?

**Comment by douglasmiller** [comment]
I think that response should be a caught as a `Recurly::Transaction::Error`. The base transaction error [extends the `UnprocessableEntity` error](https://github.com/recurly/recurly-client-ruby/blob/v2/lib/recurly/transaction/errors.rb#L27). The methods defined on the `Recurly::Transaction::Error` should capture the specifics of the error.

Since the `Recurly::Transaction::Error` extends `UnprocessableEntity`, you'd want to `rescue Recurly::Transaction::Error => e` (or one of it's sub-variants) prior to rescuing the `UnprocessableEntity` error. 

Give that a try and let us know if that works.

**Comment by marcmoy** [comment]
@douglasmiller I tried but I'm still not able to get the expected result. See anything wrong with the sample code below?
```
irb(main):051:0> e = nil
=> nil
irb(main):052:0> begin
irb(main):053:1*     transaction = Recurly::Transaction.find(ext_transaction_id)
irb(main):054:1>     invoice = transaction.invoice
irb(main):055:1>     RecurlyHelpers.refund_invoice(invoice)
irb(main):056:1> rescue Recurly::Transaction::Error => error
irb(main):057:1>     e = error
irb(main):058:1> rescue StandardError => error
irb(main):059:1>     e = error
irb(main):060:1> end
=> #<Recurly::API::UnprocessableEntity: 422 Unprocessable Entity (POST https://api.recurly.com/v2/invoices/.../refund)>
irb(main):061:0> e.class
=> Recurly::API::UnprocessableEntity
irb(main):062:0>
```

**Comment by douglasmiller** [comment]
@marcmoy, this looks like a bug in the client. I'll get this scheduled to be fixed.

**Comment by joannasese** [comment]
Hi @marcmoy. Thanks again for reporting. We've released [a patch](https://github.com/recurly/recurly-client-ruby/pull/653) that is included in [version 2.18.18](https://github.com/recurly/recurly-client-ruby/releases/tag/2.18.18) of this client library. If/when you upgrade from 2.17 to 2.18, [I don't anticipate that you'll have any issues](https://github.com/recurly/recurly-client-ruby/blob/v2/CHANGELOG.md#v2180-2019-08-21). I'll go ahead and close out this issue.
